### PR TITLE
Map all true schemas to empty object schemas in OpenAPI

### LINKED ISSUE #57041: Custom types with converters cause schema generation to fail
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Using the .NET 9 version of Microsoft.AspNetCore.OpenApi (`9.0.0-rc.1.24376.4`), in the document generation process, upon encountering a non-standard type with a custom converter that serializes into a primitive JSON value (JsonTypeInfo.Kind is None), the generator will throw an exception related to System.Text.Json's schemas.

### Expected Behavior

When a custom type with a custom converter that serializes into a JSON primitive is detected in the schema, it should have a way to represent that type as a string or number or whatever the value is supposed to be.

### Steps To Reproduce

I'm using the [Ulid](https://github.com/Cysharp/Ulid) type as an example (serializes into a string), but [here is a minimal repro](https://github.com/Auros/MinimalOpenApiCustomParameterTypes).

### Exceptions (if any)

```
System.InvalidOperationException: The node must be of type 'JsonObject'.
   at System.Text.Json.ThrowHelper.ThrowInvalidOperationException_NodeWrongType(ReadOnlySpan`1 supportedTypeNames)
   at System.Text.Json.Nodes.JsonNode.AsObject()
   at System.Text.Json.Nodes.JsonNode.set_Item(String propertyName, JsonNode value)
   at Microsoft.AspNetCore.OpenApi.JsonNodeSchemaExtensions.ApplySchemaReferenceId(JsonNode schema, JsonSchemaExporterContext context, Func`2 createSchemaReferenceId)
   at Microsoft.AspNetCore.OpenApi.OpenApiSchemaService.<>c__DisplayClass0_0.<.ctor>b__2(JsonSchemaExporterContext context, JsonNode schema)
   at System.Text.Json.Schema.JsonSchema.<ToJsonNode>g__CompleteSchema|104_0(JsonNode schema, <>c__DisplayClass104_0&)
   at System.Text.Json.Schema.JsonSchema.ToJsonNode(JsonSchemaExporterOptions options)
   at System.Text.Json.Schema.JsonSchemaExporter.GetJsonSchemaAsNode(JsonTypeInfo typeInfo, JsonSchemaExporterOptions exporterOptions)
   at System.Text.Json.Schema.JsonSchemaExporter.GetJsonSchemaAsNode(JsonSerializerOptions options, Type type, JsonSchemaExporterOptions exporterOptions)
   at Microsoft.AspNetCore.OpenApi.OpenApiSchemaService.CreateSchema(OpenApiSchemaKey key)
   at Microsoft.AspNetCore.OpenApi.OpenApiSchemaStore.GetOrAdd(OpenApiSchemaKey key, Func`2 valueFactory)
   at Microsoft.AspNetCore.OpenApi.OpenApiSchemaService.GetOrCreateSchemaAsync(Type type, ApiParameterDescription parameterDescription, Boolean captureSchemaByRef, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.OpenApi.OpenApiDocumentService.GetParametersAsync(ApiDescription description, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.OpenApi.OpenApiDocumentService.GetOperationAsync(ApiDescription description, HashSet`1 capturedTags, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.OpenApi.OpenApiDocumentService.GetOperationsAsync(IGrouping`2 descriptions, HashSet`1 capturedTags, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.OpenApi.OpenApiDocumentService.GetOpenApiPathsAsync(HashSet`1 capturedTags, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.OpenApi.OpenApiDocumentService.GetOpenApiDocumentAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Builder.OpenApiEndpointRouteBuilderExtensions.<>c__DisplayClass0_0.<<MapOpenApi>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Http.Generated.<GeneratedRouteBuilderExtensions_g>F56B68D2B55B5B7B373BA2E4796D897848BC0F04A969B1AF6260183E8B9E0BAF2__GeneratedRouteBuilderExtensionsCore.<>c__DisplayClass2_0.<<MapGet0>g__RequestHandler|5>d.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddlewareImpl.Invoke(HttpContext context)
```

### .NET Version

9.0.100-rc.1.24377.5

### Anything else?

I would have assumed a SchemaTransformer could have been used to override the type I wanted (like outlined in #56448), but it appears all transformers run after the initial document gets generated. The underlying schema generation process seems to rely on `JsonSchema`, which is both an internal API and only explicitly implements supports the common BCL types (primitives, string, Guid, Uri, etc.), so it can't really be overridden at the moment.

I personally think an API that allows setting a custom delegate (something similar but exactly `Func<Type, OpenApiSchema?>`) and having that delegate be checked first before falling back to `JsonSchema` could be an option. As far as I can tell, currently every type exposed to the schema gets run through a `converter.GetSchema()`, and if that returns null and if it's a JSON primitive, it [returns the True JsonNode](https://github.com/dotnet/runtime/blob/3471de7ffda7105f395a25a57094097a7cc0dc88/src/libraries/System.Text.Json/src/System/Text/Json/Schema/JsonSchema.cs#L143) type instead of [returning a JsonObject](https://github.com/dotnet/aspnetcore/blob/06155c05af89c957de20d2c53cee0e37171b9a09/src/OpenApi/src/Services/Schemas/OpenApiSchemaService.cs#L84), thus causing the exception).

### PR DESCRIPTION
**Pull request by captainsafia**
Closes #57041.

This PR is an extension of the changes that was introduced in https://github.com/dotnet/aspnetcore/commit/4c8b5fefd708130453d4ac029dc78e87d6fb63b7 to replace the use of `true` schemas with empty object schemas in the OpenAPI implementation.

`JsonSchemaExporter` uses the `true` schemas as catch-all schema for types. OpenAPI schema diverges from JSON schema and expects that empty schemas `{ }` be used to represent catch-all values. The previous logic special cased fixing up these mappings for `object` types. This PR updates the implementation to do the mapping if the underlying schema is a `true` schema, regardless of the type. This makes the implementation resilient to other cases where a `true` schema is returned by STJ, like [when a user defined converter is associated with a type](https://github.com/dotnet/runtime/blob/3471de7ffda7105f395a25a57094097a7cc0dc88/src/libraries/System.Text.Json/src/System/Text/Json/Schema/JsonSchemaExporter.cs#L345-L347).

With these changes, we'd expect the following diff in the generated schema.

```csharp
public class ContainerType
{
	public string Name { get; }
	public TypeWithUserDefinedConverter ChildType { get; }
}
```

```diff
{
	"type": "object",
	"properties": {
		"name": {
			"type": "string"
		},
-		"childType": true,
+		"childType": { },
	}
}
```

### CODE REVIEW

**Review by mikekistler — APPROVED** [review]
Looks good! 👍

File: src/OpenApi/src/Services/Schemas/OpenApiSchemaService.cs [hunk 1]
```diff
@@ -82,9 +82,9 @@ internal sealed class OpenApiSchemaService(
                 };
             }
             // STJ uses `true` in place of an empty object to represent a schema that matches
-            // anything. We override this default behavior here to match the style traditionally
-            // expected in OpenAPI documents.
-            if (type == typeof(object))
+            // anything (like the `object` type) or types with user-defined converters. We override
+            // this default behavior here to match the style traditionally expected in OpenAPI documents.
```
Comment (amcasey) [inline src/OpenApi/src/Services/Schemas/OpenApiSchemaService.cs:86]: What does "traditionally" mean here?  Is there a newer style of OpenAPI documents?

File: src/OpenApi/src/Services/Schemas/OpenApiSchemaService.cs [hunk 2]
```diff
@@ -82,9 +82,9 @@ internal sealed class OpenApiSchemaService(
...
+            if (schema.GetValueKind() == JsonValueKind.True)
```
Comment (amcasey) [inline src/OpenApi/src/Services/Schemas/OpenApiSchemaService.cs:87]: Idle thought: what happens for `dynamic`?

**Review by amcasey — APPROVED** [review]

File: src/OpenApi/src/Services/Schemas/OpenApiSchemaService.cs (hunk 1 above)
Comment (captainsafia) [reply src/OpenApi/src/Services/Schemas/OpenApiSchemaService.cs:86]: Yes, OpenAPI v3.1 is an upcoming version that has full parity with JSON schema and would permit `true` schemas being used as catch-all schemas. I'll update this comment to be more explicit.

File: src/OpenApi/src/Services/Schemas/OpenApiSchemaService.cs (hunk 2 above)
Comment (captainsafia) [reply src/OpenApi/src/Services/Schemas/OpenApiSchemaService.cs:87]: Good question! Looks like it gets handled similarily to `object foo`. I added test coverage for this in https://github.com/dotnet/aspnetcore/pull/57067/commits/fa39cfcfd1c32f410dc501a1e184f57539dcd144.

**Pull request merged**